jsonschema-core.md

```diff
@@ -523,11 +523,23 @@ to inform the user of the evaluation path that produced the error or annotation.
 
 ### Keyword Interactions
 
-Keyword behavior MAY be defined in terms of the annotation results of
-[subschemas](#root) and/or adjacent keywords (keywords within the same schema
-object) and their subschemas. Such keywords MUST NOT result in a circular
-dependency. Keywords MAY modify their behavior based on the presence or absence
-of another keyword in the same [schema object](#schema-document).
+Keywords MAY modify their behavior based on the presence or absence of another
```

Comment:
Do we need to be explicit that keywords need to be defined so that this interaction is one-way?  E.g. the way it's defined now, `minContains` can modify the validation result of `contains`.  (I have an item in the linked issue to address this.)

Reply:
Isn't that what the "circular dependency" clause says? I might be missing the point.

Reply:
No, by "one-way" I mean that a keyword can't be used to explicitly modify another keyword's validation result.  The logic has to be in the dependent keyword.

Reply:
I'm still not following the distinction. As originally defined, the result of `contains` depends on the annotations of `minContains` and the result of `minContains` depends on the annotations of `contains`. That sounds like the circular dependency that's mentioned. I'm not seeing what else might need to be said.

Reply:
I think I see what you're saying. It's confusing to me because the direction of operation depends on your view point. `additionalProperties` changes it's result based on the value of `properties`, but if you look at it from a different perspective, `properties` changes the result of `additionalProperties`. That is, if I add a property to `properties`, the result of `additionalProperties` changes. `contains`/`minContains` is the same. We want people to specify that `contains` depends on `minContains`, not that `minContains` changes the result of `contains` even though both perspectives are technically correct.

I think different directions of operation are just different perspectives of the same behavior. I don't think that direction is the reason `contains`/`minContains` was problematic. It was problematic because it can circular dependencies.

Assuming I'm not still missing something, this sounds like guidance for spec authors and although useful, probably doesn't belong in the spec.

Reply:
Okay...  The different perspectives thing is technically correct, I guess, but I'm asking if it's important that keywords need to be specified so that the dependent keyword has the variable behavior.

If `foo` depends on `bar` in some way, the specification shouldn't say that `bar` changes the outcome of `foo`; it should say that `foo` can change its own behavior depending on `bar`.

Reply:
> I'm asking if it's important that keywords need to be specified so that the dependent keyword has the variable behavior.

I think we should definitely stick to defining keywords that way, but that's more of spec author style guide requirement. It doesn't constraint implementations in any way, so I don't think it needs to be in the spec.

Reply:
> that's more of spec author style guide requirement

I see no reason why a spec can't contain requirements for extension specs.  I'm [doing that](https://github.com/json-schema-org/json-schema-spec/pull/1429/files#diff-5f8683f3cbc29d6dcf36b2664eda6d2efbd18a7074546d1c9105a60e239e5f51R2085-R2087) with the output PR.

In this section, Core is describing how keyword interaction works _in general_.  I think it's a good idea to set some guidelines for people who want to define their own keywords.

Reply:
I don't feel too strongly about this one, but I think it would be good to get some other opinions. My opinion is that this is an example of putting stuff in the spec that doesn't necessarily belong there because we currently have nowhere else to put that kind of content.

Reply:
> > I'm asking if it's important that keywords need to be specified so that the dependent keyword has the variable behavior.
> 
> I think we should definitely stick to defining keywords that way, but that's more of spec author style guide requirement. It doesn't constraint implementations in any way, so I don't think it needs to be in the spec.


I agree we should be defining keywords that way, and I feel that it should be in the specifciation too. The specification [defines how vocabularies are defined](https://json-schema.org/draft/2020-12/json-schema-core.html#name-schema-vocabularies). Therefore, it sets up the foundations for the JSON Schema ecosystem, not just how implementations should function. You may suggest that such content should be elsewhere, and the specification should only contain the requirements of an implementaiton, however some prose and context is helpful and relevant, which is why it's here. It doesn't make any requirements (We could if we wanted to), but some directions like this do belong in the spec IMHO.

Reply:
Relevant text is now at line 541.  Leaving it in as a soft recommendation for supplementary specs.